// @react-native/codegen/lib/generators/modules/__test_fixtures__/fixtures.js
imary
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: primaryTelexNumber
schemaIDGUID:: IcGWAtpA0RGpwAAA+ANnwQ==
attributeSecurityGUID:: hri1d0qU0RGuvQAA+ANnwQ==
systemOnly: FALSE

dn: CN=Title,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: Title
attributeID: 2.5.4.12
attributeSyntax: 2.5.5.12
isSingleValued: TRUE
rangeLower: 1
rangeUpper: 64
showInAdvancedViewOnly: TRUE
adminDisplayName: Title
adminDescription: Title
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: title
schemaIDGUID:: VXqWv+YN0BGihQCqADBJ4g==
attributeSecurityGUID:: VAGN5Pi80RGHAgDAT7lgUA==
systemOnly: FALSE

dn: CN=User-Comment,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: User-Comment
attributeID: 1.2.840.113556.1.4.156
attributeSyntax: 2.5.5.12
isSingleValued: TRUE
showInAdvancedViewOnly: TRUE
adminDisplayName: User-Comment
adminDescription: User-Comment
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: comment
schemaIDGUID:: anqWv+YN0BGihQCqADBJ4g==
attributeSecurityGUID:: Qi+6WaJ50BGQIADAT8LTzw==
systemOnly: FALSE

dn: CN=Address-Home,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: Address-Home
attributeID: 1.2.840.113556.1.2.617
attributeSyntax: 2.5.5.12
isSingleValued: TRUE
rangeLower: 1
rangeUpper: 4096
showInAdvancedViewOnly: TRUE
adminDisplayName: Address-Home
adminDescription: Address-Home
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: homePostalAddress
schemaIDGUID:: gVd3FvNH0RGpwwAA+ANnwQ==
attributeSecurityGUID:: hri1d0qU0RGuvQAA+ANnwQ==
systemOnly: FALSE

dn: CN=houseIdentifier,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: houseIdentifier
attributeID: 2.5.4.51
attributeSyntax: 2.5.5.12
isSingleValued: FALSE
rangeLower: 1
rangeUpper: 32768
showInAdvancedViewOnly: TRUE
adminDisplayName: houseIdentifier
adminDescription:
 The houseIdentifier attribute type specifies a linguistic construct used to id
 entify a particular building, for example a house number or house name relativ
 e to a street, avenue, town or city, etc.
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: houseIdentifier
schemaIDGUID:: t5hTpErEtk6C0xPBCUbb/g==
systemOnly: FALSE

dn: CN=Default-Class-Store,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: Default-Class-Store
attributeID: 1.2.840.113556.1.4.213
attributeSyntax: 2.5.5.1
isSingleValued: FALSE
showInAdvancedViewOnly: TRUE
adminDisplayName: Default-Class-Store
oMObjectClass:: KwwCh3McAIVK
adminDescription: Default-Class-Store
oMSyntax: 127
searchFlags: 0
lDAPDisplayName: defaultClassStore
schemaIDGUID:: SHmWv+YN0BGihQCqADBJ4g==
systemOnly: FALSE

dn: CN=Dynamic-LDAP-Server,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: Dynamic-LDAP-Server
attributeID: 1.2.840.113556.1.4.537
attributeSyntax: 2.5.5.1
isSingleValued: TRUE
showInAdvancedViewOnly: TRUE
adminDisplayName: Dynamic-LDAP-Server
oMObjectClass:: KwwCh3McAIVK
adminDescription: Dynamic-LDAP-Server
oMSyntax: 127
searchFlags: 0
lDAPDisplayName: dynamicLDAPServer
schemaIDGUID:: IYBFUmrK0BGv/wAA+ANnwQ==
systemOnly: FALSE

dn: CN=Preferred-OU,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: Preferred-OU
attributeID: 1.2.840.113556.1.4.97
attributeSyntax: 2.5.5.1
isSingleValued: TRUE
showInAdvancedViewOnly: TRUE
adminDisplayName: Preferred-OU
oMObjectClass:: KwwCh3McAIVK
adminDescription: Preferred-OU
oMSyntax: 127
searchFlags: 16
lDAPDisplayName: preferredOU
schemaIDGUID:: /3mWv+YN0BGihQCqADBJ4g==
systemOnly: FALSE

dn: CN=User-Parameters,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: User-Parameters
attributeID: 1.2.840.113556.1.4.138
attributeSyntax: 2.5.5.12
isSingleValued: TRUE
rangeLower: 0
rangeUpper: 32767
showInAdvancedViewOnly: TRUE
adminDisplayName: User-Parameters
adminDescription: User-Parameters
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: userParameters
schemaIDGUID:: bXqWv+YN0BGihQCqADBJ4g==
attributeSecurityGUID:: AEIWTMAg0BGnaACqAG4FKQ==
systemOnly: FALSE

dn: CN=X509-Cert,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: X509-Cert
attributeID: 2.5.4.36
attributeSyntax: 2.5.5.10
isSingleValued: FALSE
rangeUpper: 32768
showInAdvancedViewOnly: TRUE
adminDisplayName: X509-Cert
adminDescription: X509-Cert
oMSyntax: 4
searchFlags: 0
lDAPDisplayName: userCertificate
schemaIDGUID:: f3qWv+YN0BGihQCqADBJ4g==
attributeSecurityGUID:: hri1d0qU0RGuvQAA+ANnwQ==
systemOnly: FALSE
isMemberOfPartialAttributeSet: TRUE

dn: CN=audio,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: audio
attributeID: 0.9.2342.19200300.100.1.55
attributeSyntax: 2.5.5.10
isSingleValued: FALSE
rangeUpper: 250000
showInAdvancedViewOnly: FALSE
adminDisplayName: audio
adminDescription:
 The Audio attribute type allows the storing of sounds in the Directory.
oMSyntax: 4
searchFlags: 0
lDAPDisplayName: audio
schemaIDGUID:: JNLh0KDhzkKi2nk7pSRPNQ==
systemOnly: FALSE

dn: CN=carLicense,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: carLicense
attributeID: 2.16.840.1.113730.3.1.1
attributeSyntax: 2.5.5.12
isSingleValued: FALSE
showInAdvancedViewOnly: FALSE
adminDisplayName: carLicense
adminDescription: Vehicle license or registration plate.
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: carLicense
schemaIDGUID:: kpwV1H2Vh0qKZ40pNOAWSQ==
systemOnly: FALSE

dn: CN=departmentNumber,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: departmentNumber
attributeID: 2.16.840.1.113730.3.1.2
attributeSyntax: 2.5.5.12
isSingleValued: FALSE
showInAdvancedViewOnly: FALSE
adminDisplayName: departmentNumber
adminDescription: Identifies a department within an organization.
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: departmentNumber
schemaIDGUID:: 7vaevsfLIk+ye5aWfn7lhQ==
systemOnly: FALSE

dn: CN=Employee-Number,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: Employee-Number
attributeID: 1.2.840.113556.1.2.610
attributeSyntax: 2.5.5.12
isSingleValued: TRUE
rangeLower: 1
rangeUpper: 512
showInAdvancedViewOnly: TRUE
adminDisplayName: Employee-Number
adminDescription: Employee-Number
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: employeeNumber
schemaIDGUID:: 73PfqOrF0RG7ywCAx2ZwwA==
systemOnly: FALSE

dn: CN=Employee-Type,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: Employee-Type
attributeID: 1.2.840.113556.1.2.613
attributeSyntax: 2.5.5.12
isSingleValued: TRUE
rangeLower: 1
rangeUpper: 256
showInAdvancedViewOnly: TRUE
adminDisplayName: Employee-Type
adminDescription: Employee-Type
oMSyntax: 64
searchFlags: 16
lDAPDisplayName: employeeType
schemaIDGUID:: 8HPfqOrF0RG7ywCAx2ZwwA==
systemOnly: FALSE

dn: CN=jpegPhoto,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: jpegPhoto
attributeID: 0.9.2342.19200300.100.1.60
attributeSyntax: 2.5.5.10
isSingleValued: FALSE
showInAdvancedViewOnly: FALSE
adminDisplayName: jpegPhoto
adminDescription:
 Used to store one or more images of a person using the JPEG File Interchange F
 ormat [JFIF].
oMSyntax: 4
searchFlags: 0
lDAPDisplayName: jpegPhoto
schemaIDGUID:: cgXIusQJqU+a5nYo162+Dg==
systemOnly: FALSE

dn: CN=labeledURI,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: labeledURI
attributeID: 1.3.6.1.4.1.250.1.57
attributeSyntax: 2.5.5.12
isSingleValued: FALSE
showInAdvancedViewOnly: FALSE
adminDisplayName: labeledURI
adminDescription:
 A Uniform Resource Identifier followed by a label. The label is used to descri
 be the resource to which the URI points, and is intended as a friendly name fi
 t for human consumption.
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: labeledURI
schemaIDGUID:: RrtpxYDGvESic+bCJ9cbRQ==
systemOnly: FALSE

dn: CN=photo,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: photo
attributeID: 0.9.2342.19200300.100.1.7
attributeSyntax: 2.5.5.10
isSingleValued: FALSE
showInAdvancedViewOnly: FALSE
adminDisplayName: photo
adminDescription:
 An object encoded in G3 fax as explained in recommendation T.4, with an ASN.1
 wrapper to make it compatible with an X.400 BodyPart as defined in X.420.
oMSyntax: 4
searchFlags: 0
lDAPDisplayName: photo
schemaIDGUID:: aJeXnBq6CEyWMsalwe1kmg==
systemOnly: FALSE

dn: CN=preferredLanguage,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: preferredLanguage
attributeID: 2.16.840.1.113730.3.1.39
attributeSyntax: 2.5.5.12
isSingleValued: TRUE
showInAdvancedViewOnly: FALSE
adminDisplayName: preferredLanguage
adminDescription: The preferred written or spoken language for a person.
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: preferredLanguage
schemaIDGUID:: 0OBrhecY4UaPX37k2QIODQ==
systemOnly: FALSE

dn: CN=roomNumber,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: roomNumber
attributeID: 0.9.2342.19200300.100.1.6
attributeSyntax: 2.5.5.12
isSingleValued: FALSE
showInAdvancedViewOnly: FALSE
adminDisplayName: roomNumber
adminDescription: The room number of an object.
oMSyntax: 64
searchFlags: 0
lDAPDisplayName: roomNumber
schemaIDGUID:: wvjXgSfjDUqRxrQtQAkRXw==
systemOnly: FALSE

dn: CN=secretary,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: secretary
attributeID: 0.9.2342.19200300.100.1.21
attributeSyntax: 2.5.5.1
isSingleValued: FALSE
showInAdvancedViewOnly: FALSE
adminDisplayName: secretary
oMObjectClass:: KwwCh3McAIVK
adminDescription: Specifies the secretary of a person.
oMSyntax: 127
searchFlags: 0
lDAPDisplayName: secretary
schemaIDGUID:: mi0HAa2YU0qXROg+KHJ4+w==
systemOnly: FALSE

dn: CN=uid,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: uid
attributeID: 0.9.2342.19200300.100.1.1
attributeSyntax: 2.5.5.12
isSingleValued: FALSE
showInAdvancedViewOnly: FALSE
adminDisplayName: uid
adminDescription: A user ID.
oMSyntax: 64
searchFlags: 8
lDAPDisplayName: uid
schemaIDGUID:: oPywC4ken0KQGhQTiU2fWQ==
attributeSecurityGUID:: Qi+6WaJ50BGQIADAT8LTzw==
systemOnly: FALSE

dn: CN=userPKCS12,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: userPKCS12
attributeID: 2.16.840.1.113730.3.1.216
attributeSyntax: 2.5.5.10
isSingleValued: FALSE
showInAdvancedViewOnly: FALSE
adminDisplayName: userPKCS12
adminDescription:
 PKCS #12 PFX PDU for exchange of personal identity information.
oMSyntax: 4
searchFlags: 0
lDAPDisplayName: userPKCS12
schemaIDGUID:: tYqZI/hwB0CkwahKODEfmg==
systemOnly: FALSE

dn: CN=User-SMIME-Certificate,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: User-SMIME-Certificate
attributeID: 2.16.840.1.113730.3.140
attributeSyntax: 2.5.5.10
isSingleValued: FALSE
rangeUpper: 32768
showInAdvancedViewOnly: TRUE
adminDisplayName: User-SMIME-Certificate
adminDescription: User-SMIME-Certificate
oMSyntax: 4
searchFlags: 0
lDAPDisplayName: userSMIMECertificate
schemaIDGUID:: sp1q4TxA0RGpwAAA+ANnwQ==
attributeSecurityGUID:: hri1d0qU0RGuvQAA+ANnwQ==
systemOnly: FALSE
isMemberOfPartialAttributeSet: TRUE

dn: CN=x500uniqueIdentifier,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: attributeSchema
cn: x500uniqueIdentifier
attributeID: 2.5.4.45
attributeSyntax: 2.5.5.10
isSingleValued: FALSE
showInAdvancedViewOnly: FALSE
adminDisplayName: x500uniqueIdentifier
adminDescription:
 Used to distinguish between objects when a distinguished name has been reused.
   This is a different attribute type from both the "uid" and "uniqueIdentifier
 " types.
oMSyntax: 4
searchFlags: 0
lDAPDisplayName: x500uniqueIdentifier
schemaIDGUID:: H6F90D2KtkKwqnbJYr5xmg==
systemOnly: FALSE

dn: CN=Owner,CN=Schema,CN=Configuration,DC=X
changetype: ntdsSchemaAdd
objectClass: top
objectClass: attributeSchema
cn: Owner
attributeID: 2.5.4.32
attributeSyntax: 2.5.5.1
isSingleValued: TRUE
linkID: 44
showInAdvancedViewOnly: TRUE
adminDisplayName: Owner
oMObjectClass:: KwwCh3McAIVK
adminDescription: Owner
oMSyntax: 127
searchFlags: 0
lDAPDisplayName: owner
schemaIDGUID:: 83mWv+YN0BGihQCqADBJ4g==
systemOnly: FALSE

dn: CN=ms-Exch-Owner-BL,CN=Schema,CN=Configuration,DC=X
changetype: ntdsSchemaAdd
objectClass: top
objectClass: attributeSchema
cn: ms-Exch-Owner-BL
attributeID: 1.2.840.113556.1.2.104
attributeSyntax: 2.5.5.1
isSingleValued: FALSE
linkID: 45
showInAdvancedViewOnly: TRUE
adminDisplayName: ms-Exch-Owner-BL
oMObjectClass:: KwwCh3McAIVK
adminDescription: ms-Exch-Owner-BL
oMSyntax: 127
searchFlags: 0
lDAPDisplayName: ownerBL
schemaIDGUID:: 9HmWv+YN0BGihQCqADBJ4g==
systemOnly: TRUE

dn:
changetype: modify
add: schemaUpdateNow
schemaUpdateNow: 1
-

# Classes

dn: CN=Person,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: classSchema
cn: Person
subClassOf: top
governsID: 2.5.6.6
mayContain: attributeCertificateAttribute
rDNAttID: cn
showInAdvancedViewOnly: TRUE
adminDisplayName: Person
adminDescription: Person
objectClassCategory: 0
lDAPDisplayName: person
name: Person
schemaIDGUID:: p3qWv+YN0BGihQCqADBJ4g==
systemOnly: FALSE
systemPossSuperiors: organizationalUnit
systemPossSuperiors: container
systemMayContain: userPassword
systemMayContain: telephoneNumber
systemMayContain: sn
systemMayContain: serialNumber
systemMayContain: seeAlso
systemMustContain: cn
defaultSecurityDescriptor:
 D:S:
defaultHidingValue: TRUE
defaultObjectCategory: CN=Person,CN=Schema,CN=Configuration,DC=X

dn:
changetype: modify
add: schemaUpdateNow
schemaUpdateNow: 1
-

dn: CN=Organizational-Person,CN=Schema,CN=Configuration,DC=X
changetype: ntdsschemaadd
objectClass: top
objectClass: classSchema
cn: Organizational-Person
subClassOf: person
governsID: 2.5.6.7
mayContain: houseIdentifier
mayContain: homePostalAddress
rDNAttID: cn
showInAdvancedViewOnly: TRUE
adminDisplayName: Organizational-Person
adminDescription: Organizational-Person
objectClassCategory: 0
lDAPDisplayName: organizationalPerson
name: Organizational-Person
schemaIDGUID:: pHqWv+YN0BGihQCqADBJ4g==
systemOnly: FALSE
systemPossSuperiors: organizationalUnit
systemPossSuperiors: organization
systemPossSuperiors: container
systemMayContain: x121Address
systemMayContain: comment
systemMayContain: title
systemMayContain: co
systemMayContain: primaryTelexNumber
systemMayContain: telexNumber
systemMayContain: teletexTerminalIdentifier
systemMayContain: street
systemMayContain: st
systemMayContain: registeredAddress
systemMayContain: preferredDeliveryMethod
systemMayContain: postalCode
systemMayContain: postalAddress
systemMayContain: postOfficeBox
systemMayContain: thumbnailPhoto
systemMayContain: physicalDeliveryOfficeName
systemMayContain: pager
systemMayContain: otherPager
systemMayContain: otherTelephone
systemMayContain: mobile
systemMayContain: otherMobile
systemMayContain: primaryInternationalISDNNumber
systemMayContain: ipPhone
systemMayContain: otherIpPhone
systemMayContain: otherHomePhone
systemMayContain: homePhone
systemMayContain: otherFacsimileTelephoneNumber
systemMayContain: personalTitle
systemMayContain: middleName
systemMayContain: ou
systemMayContain: o
systemMayContain: manager
systemMayContai               properties: [
                      {
                        optional: true,
                        name: 'optionalNumberProperty',
                        typeAnnotation: {
                          type: 'NumberTypeAnnotation',
                        },
                      },
                      {
                        optional: true,
                        name: 'optionalArrayProperty',
                        typeAnnotation: {
                          type: 'ArrayTypeAnnotation',
                          elementType: {
                            type: 'NumberTypeAnnotation',
                          },
                        },
                      },
                      {
                        optional: true,
                        name: 'optionalObjectProperty',
                        typeAnnotation: {
                          type: 'ObjectTypeAnnotation',
                          properties: [
                            {
                              optional: false,
                              name: 'x',
                              typeAnnotation: {
                                type: 'NumberTypeAnnotation',
                              },
                            },
                            {
                              optional: false,
                              name: 'y',
                              typeAnnotation: {
                                type: 'NumberTypeAnnotation',
                              },
                            },
                          ],
                        },
                      },
                      {
                        optional: true,
                        name: 'optionalGenericObjectProperty',
                        typeAnnotation: {
                          type: 'GenericObjectTypeAnnotation',
                        },
                      },
                      {
                        optional: true,
                        name: 'optionalBooleanTypeProperty',
                        typeAnnotation: {
                          type: 'BooleanTypeAnnotation',
                        },
                      },
                    ],
                  },
                },
              ],
            },
          },
          {
            name: 'optionalMethod',
            optional: true,
            typeAnnotation: {
              type: 'NullableTypeAnnotation',
              typeAnnotation: {
                type: 'FunctionTypeAnnotation',
                returnTypeAnnotation: {
                  type: 'VoidTypeAnnotation',
                },
                params: [
                  {
                    optional: false,
                    name: 'options',
                    typeAnnotation: {
                      type: 'GenericObjectTypeAnnotation',
                    },
                  },
                  {
                    name: 'callback',
                    optional: false,
                    typeAnnotation: {
                      type: 'FunctionTypeAnnotation',
                      params: [],
                      returnTypeAnnotation: {
                        type: 'VoidTypeAnnotation',
                      },
                    },
                  },
                  {
                    name: 'extras',
                    optional: true,
                    typeAnnotation: {
                      type: 'ArrayTypeAnnotation',
                      elementType: {
                        type: 'ObjectTypeAnnotation',
                        properties: [
                          {
                            optional: false,
                            name: 'key',
                            typeAnnotation: {
                              type: 'StringTypeAnnotation',
                            },
                          },
                          {
                            optional: false,
                            name: 'value',
                            typeAnnotation: {
                              type: 'GenericObjectTypeAnnotation',
                            },
                          },
                        ],
                      },
                    },
                  },
                ],
              },
            },
          },
          {
            name: 'getArrays',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'VoidTypeAnnotation',
              },
              params: [
                {
                  optional: false,
                  name: 'options',
                  typeAnnotation: {
                    type: 'ObjectTypeAnnotation',
                    properties: [
                      {
                        optional: false,
                        name: 'arrayOfNumbers',
                        typeAnnotation: {
                          type: 'ArrayTypeAnnotation',
                          elementType: {
                            type: 'NumberTypeAnnotation',
                          },
                        },
                      },
                      {
                        optional: true,
                        name: 'optionalArrayOfNumbers',
                        typeAnnotation: {
                          type: 'ArrayTypeAnnotation',
                          elementType: {
                            type: 'NumberTypeAnnotation',
                          },
                        },
                      },
                      {
                        optional: false,
                        name: 'arrayOfStrings',
                        typeAnnotation: {
                          type: 'ArrayTypeAnnotation',
                          elementType: {
                            type: 'StringTypeAnnotation',
                          },
                        },
                      },
                      {
                        optional: true,
                        name: 'optionalArrayOfStrings',
                        typeAnnotation: {
                          type: 'ArrayTypeAnnotation',
                          elementType: {
                            type: 'StringTypeAnnotation',
                          },
                        },
                      },
                      {
                        optional: false,
                        name: 'arrayOfObjects',
                        typeAnnotation: {
                          type: 'ArrayTypeAnnotation',
                          elementType: {
                            type: 'ObjectTypeAnnotation',
                            properties: [
                              {
                                optional: false,
                                name: 'numberProperty',
                                typeAnnotation: {
                                  type: 'NumberTypeAnnotation',
                                },
                              },
                            ],
                          },
                        },
                      },
                    ],
                  },
                },
              ],
            },
          },
          {
            name: 'getNullableObject',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'NullableTypeAnnotation',
                typeAnnotation: {
                  type: 'ObjectTypeAnnotation',
                  properties: [],
                },
              },
              params: [],
            },
          },
          {
            name: 'getNullableGenericObject',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'NullableTypeAnnotation',
                typeAnnotation: {
                  type: 'GenericObjectTypeAnnotation',
                },
              },
              params: [],
            },
          },
          {
            name: 'getNullableArray',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'NullableTypeAnnotation',
                typeAnnotation: {
                  type: 'ArrayTypeAnnotation',
                },
              },
              params: [],
            },
          },
        ],
      },
      moduleName: 'SampleTurboModule',
    },
  },
};
const NATIVE_MODULES_WITH_TYPE_ALIASES = {
  modules: {
    AliasTurboModule: {
      type: 'NativeModule',
      aliasMap: {
        Options: {
          type: 'ObjectTypeAnnotation',
          properties: [
            {
              optional: false,
              name: 'offset',
              typeAnnotation: {
                type: 'ObjectTypeAnnotation',
                properties: [
                  {
                    optional: false,
                    name: 'x',
                    typeAnnotation: {
                      type: 'NumberTypeAnnotation',
                    },
                  },
                  {
                    optional: false,
                    name: 'y',
                    typeAnnotation: {
                      type: 'NumberTypeAnnotation',
                    },
                  },
                ],
              },
            },
            {
              optional: false,
              name: 'size',
              typeAnnotation: {
                type: 'ObjectTypeAnnotation',
                properties: [
                  {
                    optional: false,
                    name: 'width',
                    typeAnnotation: {
                      type: 'NumberTypeAnnotation',
                    },
                  },
                  {
                    optional: false,
                    name: 'height',
                    typeAnnotation: {
                      type: 'NumberTypeAnnotation',
                    },
                  },
                ],
              },
            },
            {
              optional: true,
              name: 'displaySize',
              typeAnnotation: {
                type: 'ObjectTypeAnnotation',
                properties: [
                  {
                    optional: false,
                    name: 'width',
                    typeAnnotation: {
                      type: 'NumberTypeAnnotation',
                    },
                  },
                  {
                    optional: false,
                    name: 'height',
                    typeAnnotation: {
                      type: 'NumberTypeAnnotation',
                    },
                  },
                ],
              },
            },
            {
              optional: true,
              name: 'resizeMode',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'allowExternalStorage',
              typeAnnotation: {
                type: 'BooleanTypeAnnotation',
              },
            },
          ],
        },
      },
      enumMap: {},
      spec: {
        properties: [
          {
            name: 'getConstants',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'ObjectTypeAnnotation',
                properties: [],
              },
              params: [],
            },
          },
          {
            name: 'cropImage',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'VoidTypeAnnotation',
              },
              params: [
                {
                  optional: false,
                  name: 'cropData',
                  typeAnnotation: {
                    type: 'TypeAliasTypeAnnotation',
                    name: 'Options',
                  },
                },
              ],
            },
          },
        ],
      },
      moduleName: 'AliasTurboModule',
    },
  },
};
const REAL_MODULE_EXAMPLE = {
  modules: {
    NativeCameraRollManager: {
      type: 'NativeModule',
      aliasMap: {
        PhotoIdentifierImage: {
          type: 'ObjectTypeAnnotation',
          properties: [
            {
              optional: false,
              name: 'uri',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'playableDuration',
              typeAnnotation: {
                type: 'NumberTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'width',
              typeAnnotation: {
                type: 'NumberTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'height',
              typeAnnotation: {
                type: 'NumberTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'isStored',
              typeAnnotation: {
                type: 'BooleanTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'filename',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
          ],
        },
        PhotoIdentifier: {
          type: 'ObjectTypeAnnotation',
          properties: [
            {
              optional: false,
              name: 'node',
              typeAnnotation: {
                type: 'ObjectTypeAnnotation',
                properties: [
                  {
                    optional: false,
                    name: 'image',
                    typeAnnotation: {
                      type: 'TypeAliasTypeAnnotation',
                      name: 'PhotoIdentifierImage',
                    },
                  },
                  {
                    optional: false,
                    name: 'type',
                    typeAnnotation: {
                      type: 'StringTypeAnnotation',
                    },
                  },
                  {
                    optional: false,
                    name: 'group_name',
                    typeAnnotation: {
                      type: 'StringTypeAnnotation',
                    },
                  },
                  {
                    optional: false,
                    name: 'timestamp',
                    typeAnnotation: {
                      type: 'NumberTypeAnnotation',
                    },
                  },
                  {
                    optional: false,
                    name: 'location',
                    typeAnnotation: {
                      type: 'ObjectTypeAnnotation',
                      properties: [
                        {
                          optional: false,
                          name: 'longitude',
                          typeAnnotation: {
                            type: 'NumberTypeAnnotation',
                          },
                        },
                        {
                          optional: false,
                          name: 'latitude',
                          typeAnnotation: {
                            type: 'NumberTypeAnnotation',
                          },
                        },
                        {
                          optional: true,
                          name: 'altitude',
                          typeAnnotation: {
                            type: 'NumberTypeAnnotation',
                          },
                        },
                        {
                          optional: true,
                          name: 'heading',
                          typeAnnotation: {
                            type: 'NumberTypeAnnotation',
                          },
                        },
                        {
                          optional: true,
                          name: 'speed',
                          typeAnnotation: {
                            type: 'NumberTypeAnnotation',
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            },
          ],
        },
        PhotoIdentifiersPage: {
          type: 'ObjectTypeAnnotation',
          properties: [
            {
              optional: false,
              name: 'edges',
              typeAnnotation: {
                type: 'ArrayTypeAnnotation',
                elementType: {
                  type: 'TypeAliasTypeAnnotation',
                  name: 'PhotoIdentifier',
                },
              },
            },
            {
              optional: false,
              name: 'page_info',
              typeAnnotation: {
                type: 'ObjectTypeAnnotation',
                properties: [
                  {
                    optional: false,
                    name: 'has_next_page',
                    typeAnnotation: {
                      type: 'BooleanTypeAnnotation',
                    },
                  },
                  {
                    optional: true,
                    name: 'start_cursor',
                    typeAnnotation: {
                      type: 'StringTypeAnnotation',
                    },
                  },
                  {
                    optional: true,
                    name: 'end_cursor',
                    typeAnnotation: {
                      type: 'StringTypeAnnotation',
                    },
                  },
                ],
              },
            },
          ],
        },
        GetPhotosParams: {
          type: 'ObjectTypeAnnotation',
          properties: [
            {
              optional: false,
              name: 'first',
              typeAnnotation: {
                type: 'NumberTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'after',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'groupName',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'groupTypes',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'assetType',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'maxSize',
              typeAnnotation: {
                type: 'NumberTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'mimeTypes',
              typeAnnotation: {
                type: 'ArrayTypeAnnotation',
                elementType: {
                  type: 'StringTypeAnnotation',
                },
              },
            },
          ],
        },
      },
      enumMap: {},
      spec: {
        properties: [
          {
            name: 'getConstants',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'ObjectTypeAnnotation',
                properties: [],
              },
              params: [],
            },
          },
          {
            name: 'getPhotos',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'PromiseTypeAnnotation',
              },
              params: [
                {
                  optional: false,
                  name: 'params',
                  typeAnnotation: {
                    type: 'TypeAliasTypeAnnotation',
                    name: 'GetPhotosParams',
                  },
                },
              ],
            },
          },
          {
            name: 'saveToCameraRoll',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'PromiseTypeAnnotation',
              },
              params: [
                {
                  optional: false,
                  name: 'uri',
                  typeAnnotation: {
                    type: 'StringTypeAnnotation',
                  },
                },
                {
                  optional: false,
                  name: 'type',
                  typeAnnotation: {
                    type: 'StringTypeAnnotation',
                  },
                },
              ],
            },
          },
          {
            name: 'deletePhotos',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'PromiseTypeAnnotation',
              },
              params: [
                {
                  name: 'assets',
                  optional: false,
                  typeAnnotation: {
                    type: 'ArrayTypeAnnotation',
                    elementType: {
                      type: 'StringTypeAnnotation',
                    },
                  },
                },
              ],
            },
          },
        ],
      },
      moduleName: 'CameraRollManager',
    },
    NativeExceptionsManager: {
      type: 'NativeModule',
      aliasMap: {
        StackFrame: {
          properties: [
            {
              optional: true,
              name: 'column',
              typeAnnotation: {
                type: 'NumberTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'file',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'lineNumber',
              typeAnnotation: {
                type: 'NumberTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'methodName',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'collapse',
              typeAnnotation: {
                type: 'BooleanTypeAnnotation',
              },
            },
          ],
          type: 'ObjectTypeAnnotation',
        },
        ExceptionData: {
          properties: [
            {
              optional: false,
              name: 'message',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'originalMessage',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'name',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'componentStack',
              typeAnnotation: {
                type: 'StringTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'stack',
              typeAnnotation: {
                type: 'ArrayTypeAnnotation',
                elementType: {
                  type: 'TypeAliasTypeAnnotation',
                  name: 'StackFrame',
                },
              },
            },
            {
              optional: false,
              name: 'id',
              typeAnnotation: {
                type: 'NumberTypeAnnotation',
              },
            },
            {
              optional: false,
              name: 'isFatal',
              typeAnnotation: {
                type: 'BooleanTypeAnnotation',
              },
            },
            {
              optional: true,
              name: 'extraData',
              typeAnnotation: {
                type: 'GenericObjectTypeAnnotation',
              },
            },
          ],
          type: 'ObjectTypeAnnotation',
        },
      },
      enumMap: {},
      spec: {
        properties: [
          {
            name: 'reportFatalException',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'VoidTypeAnnotation',
              },
              params: [
                {
                  optional: false,
                  name: 'message',
                  typeAnnotation: {
                    type: 'StringTypeAnnotation',
                  },
                },
                {
                  name: 'stack',
                  optional: false,
                  typeAnnotation: {
                    type: 'ArrayTypeAnnotation',
                    elementType: {
                      type: 'TypeAliasTypeAnnotation',
                      name: 'StackFrame',
                    },
                  },
                },
                {
                  optional: false,
                  name: 'exceptionId',
                  typeAnnotation: {
                    type: 'NumberTypeAnnotation',
                  },
                },
              ],
            },
          },
          {
            name: 'reportSoftException',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'VoidTypeAnnotation',
              },
              params: [
                {
                  optional: false,
                  name: 'message',
                  typeAnnotation: {
                    type: 'StringTypeAnnotation',
                  },
                },
                {
                  name: 'stack',
                  optional: false,
                  typeAnnotation: {
                    type: 'ArrayTypeAnnotation',
                    elementType: {
                      type: 'TypeAliasTypeAnnotation',
                      name: 'StackFrame',
                    },
                  },
                },
                {
                  optional: false,
                  name: 'exceptionId',
                  typeAnnotation: {
                    type: 'NumberTypeAnnotation',
                  },
                },
              ],
            },
          },
          {
            name: 'reportException',
            optional: true,
            typeAnnotation: {
              type: 'NullableTypeAnnotation',
              typeAnnotation: {
                type: 'FunctionTypeAnnotation',
                returnTypeAnnotation: {
                  type: 'VoidTypeAnnotation',
                },
                params: [
                  {
                    optional: false,
                    name: 'data',
                    typeAnnotation: {
                      type: 'TypeAliasTypeAnnotation',
                      name: 'ExceptionData',
                    },
                  },
                ],
              },
            },
          },
          {
            name: 'updateExceptionMessage',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'VoidTypeAnnotation',
              },
              params: [
                {
                  optional: false,
                  name: 'message',
                  typeAnnotation: {
                    type: 'StringTypeAnnotation',
                  },
                },
                {
                  name: 'stack',
                  optional: false,
                  typeAnnotation: {
                    type: 'ArrayTypeAnnotation',
                    elementType: {
                      type: 'TypeAliasTypeAnnotation',
                      name: 'StackFrame',
                    },
                  },
                },
                {
                  optional: false,
                  name: 'exceptionId',
                  typeAnnotation: {
                    type: 'NumberTypeAnnotation',
                  },
                },
              ],
            },
          },
          {
            name: 'dismissRedbox',
            optional: true,
            typeAnnotation: {
              type: 'NullableTypeAnnotation',
              typeAnnotation: {
                type: 'FunctionTypeAnnotation',
                returnTypeAnnotation: {
                  type: 'VoidTypeAnnotation',
                },
                params: [],
              },
            },
          },
        ],
      },
      moduleName: 'ExceptionsManager',
    },
  },
};
const CXX_ONLY_NATIVE_MODULES = {
  modules: {
    NativeSampleTurboModule: {
      type: 'NativeModule',
      aliasMap: {
        ObjectAlias: {
          type: 'ObjectTypeAnnotation',
          properties: [
            {
              name: 'x',
              optional: false,
              typeAnnotation: {
                type: 'NumberTypeAnnotation',
              },
            },
          ],
        },
      },
      enumMap: {
        NumEnum: {
          type: 'EnumDeclarationWithMembers',
          name: 'NumEnum',
          memberType: 'NumberTypeAnnotation',
          members: [
            {
              name: 'ONE',
              value: '1',
            },
            {
              name: 'TWO',
              value: '2',
            },
          ],
        },
        FloatEnum: {
          type: 'EnumDeclarationWithMembers',
          name: 'FloatEnum',
          memberType: 'NumberTypeAnnotation',
          members: [
            {
              name: 'POINT_ZERO',
              value: '0.0',
            },
            {
              name: 'POINT_ONE',
              value: '0.1',
            },
            {
              name: 'POINT_TWO',
              value: '0.2',
            },
          ],
        },
        StringEnum: {
          type: 'EnumDeclarationWithMembers',
          name: 'StringEnum',
          memberType: 'StringTypeAnnotation',
          members: [
            {
              name: 'HELLO',
              value: 'hello',
            },
            {
              name: 'GoodBye',
              value: 'goodbye',
            },
          ],
        },
      },
      spec: {
        properties: [
          {
            name: 'getMixed',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'MixedTypeAnnotation',
              },
              params: [
                {
                  name: 'arg',
                  optional: false,
                  typeAnnotation: {
                    type: 'MixedTypeAnnotation',
                  },
                },
              ],
            },
          },
          {
            name: 'getNullableNumberFromNullableAlias',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'NullableTypeAnnotation',
                typeAnnotation: {
                  type: 'NumberTypeAnnotation',
                },
              },
              params: [
                {
                  name: 'a',
                  optional: false,
                  typeAnnotation: {
                    type: 'NullableTypeAnnotation',
                    typeAnnotation: {
                      type: 'TypeAliasTypeAnnotation',
                      name: 'ObjectAlias',
                    },
                  },
                },
              ],
            },
          },
          {
            name: 'getEnums',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'StringTypeAnnotation',
              },
              params: [
                {
                  name: 'enumInt',
                  optional: false,
                  typeAnnotation: {
                    name: 'NumEnum',
                    type: 'EnumDeclaration',
                    memberType: 'NumberTypeAnnotation',
                  },
                },
                {
                  name: 'enumFloat',
                  optional: false,
                  typeAnnotation: {
                    name: 'FloatEnum',
                    type: 'EnumDeclaration',
                    memberType: 'NumberTypeAnnotation',
                  },
                },
                {
                  name: 'enumString',
                  optional: false,
                  typeAnnotation: {
                    name: 'StringEnum',
                    type: 'EnumDeclaration',
                    memberType: 'StringTypeAnnotation',
                  },
                },
              ],
            },
          },
          {
            name: 'getUnion',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'UnionTypeAnnotation',
                memberType: 'ObjectTypeAnnotation',
              },
              params: [
                {
                  name: 'chooseInt',
                  optional: false,
                  typeAnnotation: {
                    type: 'UnionTypeAnnotation',
                    memberType: 'NumberTypeAnnotation',
                  },
                },
                {
                  name: 'chooseFloat',
                  optional: false,
                  typeAnnotation: {
                    type: 'UnionTypeAnnotation',
                    memberType: 'NumberTypeAnnotation',
                  },
                },
                {
                  name: 'chooseObject',
                  optional: false,
                  typeAnnotation: {
                    type: 'UnionTypeAnnotation',
                    memberType: 'ObjectTypeAnnotation',
                  },
                },
                {
                  name: 'chooseString',
                  optional: false,
                  typeAnnotation: {
                    type: 'UnionTypeAnnotation',
                    memberType: 'StringTypeAnnotation',
                  },
                },
              ],
            },
          },
          {
            name: 'getEnumReturn',
            optional: false,
            typeAnnotation: {
              type: 'FunctionTypeAnnotation',
              returnTypeAnnotation: {
                type: 'EnumDeclaration',
                name: 'NumEnum',
                memberType: 'NumberTypeAnnotation',
              },
              params: [],
            },
          },
        ],
      },
      moduleName: 'SampleTurboModuleCxx',
      excludedPlatforms: ['iOS', 'android'],
    },
  },
};
const SAMPLE_WITH_UPPERCASE_NAME = {
  modules: {
    NativeSampleTurboModule: {
      type: 'NativeModule',
      enumMap: {},
      aliasMap: {},
      spec: {
        properties: [],
      },
      moduleName: 'SampleTurboModule',
    },
  },
};
module.exports = {
  complex_objects: COMPLEX_OBJECTS,
  two_modules_different_files: TWO_MODULES_DIFFERENT_FILES,
  empty_native_modules: EMPTY_NATIVE_MODULES,
  simple_native_modules: SIMPLE_NATIVE_MODULES,
  native_modules_with_type_aliases: NATIVE_MODULES_WITH_TYPE_ALIASES,
  real_module_example: REAL_MODULE_EXAMPLE,
  cxx_only_native_modules: CXX_ONLY_NATIVE_MODULES,
  SampleWithUppercaseName: SAMPLE_WITH_UPPERCASE_NAME,
};
